test/jit/test_hooks_modules.py

```diff
@@ -0,0 +1,531 @@
+import torch
+from typing import List, Tuple
+
+
+class SubmoduleNoForwardInputs(torch.nn.Module):
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def forward(self):
+        assert self.name == "inner_mod_name"
+
+
+class ModuleNoForwardInputs(torch.nn.Module):
+    def __init__(self, name: str, submodule_name: str):
+        super().__init__()
+        self.name = name
+        self.submodule = SubmoduleNoForwardInputs(submodule_name)
+
+    def forward(self):
+        self.submodule()
+
+
+class SubmoduleForwardSingleInput(torch.nn.Module):
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def foo(self, input: str):
+        return input
+
+    def forward(self, input: str):
+        input = input + "_inner_mod"
+        input = self.foo(input)
+        return input
+
+
+class ModuleForwardSingleInput(torch.nn.Module):
+    def __init__(self, name: str, submodule_name: str):
+        super().__init__()
+        self.name = name
+        self.submodule = SubmoduleForwardSingleInput(submodule_name)
+
+    def forward(self, input: str):
+        input = input + "_outermod"
+        return self.submodule(input)
+
+
+class ModuleDirectFowardSubmodCall(torch.nn.Module):
+    def __init__(self, name: str, submodule_name: str):
+        super().__init__()
+        self.name = name
+        self.submodule = SubmoduleForwardSingleInput(submodule_name)
+
+    def forward(self, input: str):
+        input = input + "_outermod"
+        return self.submodule.forward(input)
+
+
+class SuboduleForwardMultipleInputs(torch.nn.Module):
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def forward(self, input1: List[str], input2: str):
+        input1.append(self.name)
+        output2 = input2 + "_"
+        return input1, output2
+
+
+class ModuleForwardMultipleInputs(torch.nn.Module):
+    def __init__(self, name: str, submodule_name: str):
+        super().__init__()
+        self.name = name
+        self.submodule = SuboduleForwardMultipleInputs(submodule_name)
+
+    def forward(self, input1: List[str], input2: str):
+        input1.append(self.name)
+        return self.submodule(input1, input2)
+
+
+class SubmoduleForwardTupleInput(torch.nn.Module):
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def forward(self, input: Tuple[int]):
+        input_access = input[0]
+        return (1,)
+
+
+class ModuleForwardTupleInput(torch.nn.Module):
+    def __init__(self, name: str, submodule_name: str):
+        super().__init__()
+        self.name = name
+        self.submodule = SubmoduleForwardTupleInput(submodule_name)
+
+    def forward(self, input: Tuple[int]):
+        input_access = input[0]
+        return self.submodule((1,))
+
+
+# Tests for JIT forward hooks and pre-hooks
+def test_module_no_forward_input_model():
```

Comment:
Nit: Usually `test_` is used for test methods of a class that extends `unittest.TestCase` and it might confuse people to see such a prefix on a function that is not a unit test. Consider renaming these to something like `create_module_no_forward_input()`.